default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1602 @@
+                     if part not in self.strictly_worse_parts]
+        pid = self.production_cost.keys()[0]  # TODO: Check if actually location invariant
+        for new_part in new_parts:
+            self.strictly_worse_parts[new_part.name] = []
+            for part_class in ALL_META_CLASSES:
+                if new_part.partClass in part_class:
+                    for old_part in [_get_part_type(part) for part in self.strictly_worse_parts
+                                     if part != new_part.name]:
+                        if old_part.partClass in part_class:
+                            if new_part.capacity >= old_part.capacity:
+                                a = new_part
+                                b = old_part
+                            else:
+                                a = old_part
+                                b = new_part
+                            if (self.production_cost[pid][a.name] <= self.production_cost[pid][b.name]
+                                    and {x for x in a.mountableSlotTypes} >= {x for x in b.mountableSlotTypes}):
+                                # ToDo: Maybe add production_time as additional condition?
+                                self.strictly_worse_parts[a.name].append(b.name)
+                    break
+        available_parts = sorted(self.strictly_worse_parts.keys(),
+                                 key=lambda item: _get_part_type(item).capacity, reverse=True)
+
+        # in case of a load, we need to rebuild our Cache.
+        if not self.testhulls:
+            print "Testhull cache not found. This may happen only at first turn after game start or load."
+            for hullname in available_hulls:
+                des = [des for des in testdesign_names_part if des.endswith(hullname)]
+                if des:
+                    self.testhulls.add(hullname)
+            if verbose:
+                print "Rebuilt Cache. The following hulls are used in testdesigns for parts: ", self.testhulls
+
+        # 1. Update hull test designs
+        print "Updating Testdesigns for hulls..."
+        if verbose:
+            print "Available Hulls: ", available_hulls
+            print "Existing Designs (prefix: %s): " % TESTDESIGN_NAME_HULL,
+            print [x.replace(TESTDESIGN_NAME_HULL, "") for x in testdesign_names_hull]
+        for hull in [get_hulltype(hullname) for hullname in available_hulls
+                     if "%s_%s" % (TESTDESIGN_NAME_HULL, hullname) not in testdesign_names_hull]:
+            num_slots = len([slot for slot in hull.slots])
+            partlist = num_slots * [""]
+            testdesign_name = "%s_%s" % (TESTDESIGN_NAME_HULL, hull.name)
+            try:
+                lcopy = list(partlist)
+                res = fo.issueCreateShipDesignOrder(testdesign_name, "TESTPURPOSE ONLY", hull.name,
+                                                    lcopy, "", "fighter", False)
+                del lcopy
```

Comment:
I just looked into this a little more and found a  pretty good discussion here: http://stackoverflow.com/a/17684703; it also includes some info that may be more helpful in the future regarding interaction with C threads.  The key take away for this discussion would be that when the C++ code explicitly accepts the argument as a  boost::python::list (which is the case here) then the boost wrapper for this can be considered to act as a kind of smartpointer that interacts just fine with the python garbage collector.  And if the C++ code did not take it as a boost::python object then boost would have made a copy to be dealt with on the C++ side and again no issue for the python garbage collector.  So it seems the explicit "del lcopy" is not needed.  It also seems to me that Cj is right about a copy not being needed in this particular case.